Cython/Utility/FunctionArguments.c

```diff
@@ -241,15 +241,43 @@ static int __Pyx_ParseOptionalKeywords(
     int kwds_is_tuple = CYTHON_METH_FASTCALL && likely(PyTuple_Check(kwds));
 
     while (1) {
+#if CYTHON_AVOID_BORROWED_REFS
+        Py_XDECREF(key); key=0;
+        Py_XDECREF(value); value=0;
+#endif
         if (kwds_is_tuple) {
-            if (pos >= PyTuple_GET_SIZE(kwds)) break;
+            Py_ssize_t size;
+#if CYTHON_ASSUME_SAFE_MACROS
+            size = PyTuple_GET_SIZE(kwds);
+#else
+            size = PyTuple_Size(kwds);
+            if (size < 0) goto bad;
+#endif
+            if (pos >= size) break;
+#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
             key = PyTuple_GET_ITEM(kwds, pos);
+#elif !CYTHON_AVOID_BORROWED_REFS
+            key = PyTuple_GetItem(kwds, pos);
+#else
+            key = __Pyx_PySequence_ITEM(kwds, pos);
+#endif
```

Comment:
Does this look any better if you reverse the conditions to put `#if AVOID_BORROWED_REFS` first and `#if SAFE_MACROS` second?